source/common/http/filter_manager.cc

```diff
@@ -880,16 +880,13 @@ void FilterManager::sendDirectLocalReply(
             state_.non_100_response_headers_encoded_ = true;
             filter_manager_callbacks_.encodeHeaders(*filter_manager_callbacks_.responseHeaders(),
                                                     end_stream);
-            maybeEndEncode(end_stream);
+            maybeEndEncode(end_stream && !state_.local_complete_);
```

Comment:
Sorry where is local_complete_ set to true? Before the filter chain starts iterating? This is not super intuitive. Can you add more comments?

Reply:
but why do we need to sometimes end encode after encode headers / encode data, and sometimes outside of the EncodeFunctions?  Can't we just remove the one at the end, or remove these two calls?

Reply:
I looked through the code and I think it was there because the call to encode headers could trigger a reset, in which case encodeData wouldn't be called, so in that case endStream wouldn't be called. Calling reset would call local_complete_, and so endStream would be called at the end of the function.

That said, I tried removing it and no integration tests seem to fail. Seems like `doEndStream` is called as part of reset anyways, so it might not be necessary? Pushed this change, so let's see if CI passes